# vivaldi
  <!-- badges: start -->
  [![R-CMD-check](https://github.com/GreshamLab/vivaldi/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/GreshamLab/vivaldi/actions/workflows/R-CMD-check.yaml)
  <!-- badges: end -->

*vivaldi* (**Vi**ral **Va**riant **L**ocation **a**nd **D**iversity) is an R package for analyzing viral genomes using next generation DNA sequencing.

The standard input is a vcf file.

To install this package:

## Installation

``` r
install.packages("vivaldi")
```

### Development version

To get a bug fix or to use a feature from the development version, you
can install the development version of vivaldi from GitHub.

``` r
# install.packages("devtools")
devtools::install_github("GreshamLab/vivaldi")
```

Authors: Marrisa Knoll, Katherine Johnson, Megan Hockman, Eric Borenstein, Mohammed Khalfan, Elodie Ghedin, David Gresham

Email: <dg107@nyu.edu>

------------------------------------------------------------------------

If you use vivaldi in published research, please cite:

Roder, A. E., et al. "Optimized Quantification of Intrahost Viral Diversity in SARS-CoV-2 and Influenza Virus Sequence Data." bioRxiv (2022): 2021-05
